### PR TITLE
Raise error on to_code when entity didn't contain __code__ attr

### PR DESCRIPTION
**Pull request by ilhamfp**
This PR address #30895 issue.

### CODE REVIEW

File: tensorflow/python/autograph/impl/api.py [hunk 1]
```diff
@@ -721,6 +721,10 @@ def to_code_v1(entity,
   del arg_values
   del arg_types
   del indentation
+  if not hasattr(type(entity), '__code__'):
```
Comment (mdanatg) [inline tensorflow/python/autograph/impl/api.py]: I fear performing this check here is too early - these is logic to unpack partials, method-wrappers, and others. A better place to raise the error is just before accessing the __code__ field.

**Review by mdanatg — CHANGES_REQUESTED** [review]

File: tensorflow/python/autograph/impl/api.py (hunk 1 above)
Comment (ilhamfp) [reply tensorflow/python/autograph/impl/api.py]: Okay. So just delete this one and leave the other one on the bottom?

Comment (mdanatg) [reply tensorflow/python/autograph/impl/api.py]: Rather, adding it in conversion.py at line 310, just before it tries to make the offending call, would be better. At that point, we can't tell whether the caller is to_code, to_graph or something else. So the message should be more generic a bit:

"[...]. If this is a @tf.function, try passing f.python_function instead."

Comment (ilhamfp) [reply tensorflow/python/autograph/impl/api.py]: Ah I see. So the corrent place to put it is [here](https://github.com/tensorflow/tensorflow/blob/63fdf5ffe374e408d85ac0075fc51b8677a29f52/tensorflow/python/util/tf_inspect.py#L345)?  

Like this?
```
  if not hasattr(type(tf_decorator.unwrap(object)[1]), '__code__'):
    raise ValueError('Cannot apply autograph to a function that doesn\'t '
                     'expose a __code__ object. If this is a @tf.function,'
                     ' please use to_code(f.python_function) instead.')
```

Comment (ilhamfp) [reply tensorflow/python/autograph/impl/api.py]: Alright, thank you!

File: tensorflow/python/autograph/impl/api_test.py [hunk 2]
```diff
@@ -921,6 +921,16 @@ def test_fn(x, s):
...
 
+  def test_to_code_with_wrapped_function(self):
+
+    @def_function.function
+    def test_fn(x, s):
+      while tf.reduce_sum(x) > s:
+        x /= 2
+      return x
+
+    self.assertRaises(ValueError, api.to_code(test_fn))
```
Comment (mdanatg) [inline tensorflow/python/autograph/impl/api_test.py]: I think it's a good idea to check the message a bit as well:

```
with assertRaisesRegex(ValueError, 'try passing.*python_function'):
  api.to_code(test_fn)
```

**Review by mdanatg — DISMISSED** [review]

File: tensorflow/python/autograph/impl/api_test.py (hunk 2 above)
Comment (ilhamfp) [reply tensorflow/python/autograph/impl/api_test.py]: Okay. I will change it to just that because it is more specific.

**Review by mdanatg — DISMISSED** [review]

**Review by alextp — DISMISSED** [review]

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by kkimdev** [comment]
@ilhamfp Could you look into the failed test?  You can click "Details" next to each test, e.g., "Windows Bezel" and see the details, e.g., https://source.cloud.google.com/results/invocations/38d5d46d-ad98-4052-b26c-557478091394/targets/%2F%2Ftensorflow%2Fpython%2Fautograph%2Fimpl:api_test/tests

We have to make sure that all the tests are passing before we land.  I didn't take a look those failing tests closely yet, but please let me know in case you need any help!

**Comment by ilhamfp** [comment]
@kkimdev Ah, I think I've found the error. I should not check for ValueError since any error is converted into ConversionError [here](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/autograph/impl/api.py#L607-L618). I've changed it to `self.assertRaisesRegex(Exception, 'try passing.*python_function')`.

**Comment by gbaned** [comment]
Changes have been merged into master by commit 975b093. So closing the PR.

**Pull request closed without merging**